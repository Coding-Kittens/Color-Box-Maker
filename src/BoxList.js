import Box from "./Box"
import NewBoxForm from './NewBoxForm'
import {useState} from 'react';

const BoxList = ()=>{

const [boxes,setBoxes] = useState(null);
const [newId,setNewId] = useState(0);


const handleSubmit =(event,newBox)=>{
  event.preventDefault();
  newBox.id = newId;
  setNewId((n)=> n = newId+1);
  boxes? setBoxes((n)=> n = [...boxes,newBox]): setBoxes(()=> [newBox]);
}

const deleteBox =(event)=>{
setBoxes((n)=> n = boxes.filter(box=> box.id != event.target.id));
}


return <>
<NewBoxForm handleSubmit={handleSubmit}/>
{boxes? <ul>{boxes.map(box=><Box key={box.id} id={box.id} height={box.height} width={box.width} backgroundColor={box.color} deleteBox={deleteBox}/>)}</ul> :null}
</>

}
export default BoxList;
