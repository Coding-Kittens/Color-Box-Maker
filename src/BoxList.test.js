import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

it("should renders without crashing", () => {
  render(<BoxList />);
});

it("should match snapshot", () => {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

describe("Logic tests", () => {
  let boxList;
  let width;
  let height;
  let color;
  let submitBtn;

  beforeEach(() => {
    boxList = render(<BoxList />);
    width = boxList.getByLabelText("Width");
    height = boxList.getByLabelText("Height");
    color = boxList.getByLabelText("Color");
    submitBtn = boxList.queryByText("Add Box");
  });

  it("should be able to edit the form and should clear the input values when the form is submitted", () => {
    //the form should be empty
    expect(width.value).toEqual("");
    expect(height.value).toEqual("");
    expect(color.value).toEqual("#000000");

    //fill out the form
    fireEvent.change(width, { target: { value: "50px" } });
    fireEvent.change(height, { target: { value: "40px" } });
    fireEvent.change(color, { target: { value: "#497e94" } });

    //the form input values should be correct
    expect(width.value).toEqual("50px");
    expect(height.value).toEqual("40px");
    expect(color.value).toEqual("#497e94");

    fireEvent.click(submitBtn);

    //the form should be empty after submit
    expect(width.value).toEqual("");
    expect(height.value).toEqual("");
    expect(color.value).toEqual("#000000");
  });

  it("should be able to make a new box", () => {
    //there should not be a box
    expect(boxList.container.querySelector(".Box")).not.toBeInTheDocument();

    //make a box
    fireEvent.change(width, { target: { value: "50px" } });
    fireEvent.change(height, { target: { value: "50px" } });
    fireEvent.change(color, { target: { value: "#497e94" } });
    fireEvent.click(submitBtn);

    //there should be a new box
    expect(boxList.container.querySelector(".Box")).toBeInTheDocument();
  });

  it("should delete the box", () => {
    //make a box
    fireEvent.change(width, { target: { value: "50px" } });
    fireEvent.change(height, { target: { value: "50px" } });
    fireEvent.change(color, { target: { value: "#497e94" } });
    fireEvent.click(submitBtn);

    //there should be a new box
    expect(boxList.container.querySelector(".Box")).toBeInTheDocument();

    //delete the box
    const deleteBtn = boxList.queryByText("X");
    fireEvent.click(deleteBtn);


      //there should nolonger be a box
    expect(boxList.container.querySelector(".Box")).not.toBeInTheDocument();
  });
});
