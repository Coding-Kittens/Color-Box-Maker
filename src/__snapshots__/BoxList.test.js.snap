// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="NewBoxForm"
  >
    <label>
      Width
      <input
        name="width"
        placeholder="100px"
        type="text"
        value=""
      />
    </label>
    <label>
      Height
      <input
        name="height"
        placeholder="30%"
        type="text"
        value=""
      />
    </label>
    <label>
      Color
      <input
        name="color"
        type="color"
        value="#000000"
      />
    </label>
    <button
      type="submit"
    >
      Add Box
    </button>
  </form>
</DocumentFragment>
`;
